picasso/src/main/java/com/squareup/picasso3/BitmapHunter.kt

```diff
@@ -0,0 +1,350 @@
+        picasso = picasso,
+        key = action.request.key,
+        data = action.request,
+        dispatcher = dispatcher,
+        cache = cache,
+        requestHandler = ERRORING_HANDLER,
+        retryCount = ERRORING_HANDLER.retryCount,
+        action = action,
+        priority = action.request.priority
+      )
+    }
+
+    fun updateThreadName(data: Request) {
+      val name = data.name
+      val builder = NAME_BUILDER.get()?.also {
+        it.ensureCapacity(Utils.THREAD_PREFIX.length + name.length)
+        it.replace(Utils.THREAD_PREFIX.length, it.length, name)
+      } ?: return
+
+      Thread.currentThread().name = builder.toString()
+    }
+
+    fun applyTransformations(
+      picasso: Picasso,
+      data: Request,
+      transformations: List<Transformation>,
+      result: RequestHandler.Result.Bitmap
+    ): RequestHandler.Result.Bitmap? {
+      var res = result
+
+      for (i in 0..transformations.lastIndex) {
+        val transformation = transformations[i]
+        val newResult = try {
+          val transformedResult = transformation.transform(res)
+          if (picasso.loggingEnabled) {
+            Utils.log(Utils.OWNER_HUNTER, Utils.VERB_TRANSFORMED, data.logId(), "from transformations")
+          }
+
+          transformedResult
+        } catch (e: RuntimeException) {
+          Picasso.HANDLER.post {
+            throw RuntimeException("Transformation ${transformation.key()} crashed with exception.", e)
+          }
+
+          return null
+        }
+
+        val bitmap = newResult.bitmap
+        if (bitmap.isRecycled) {
+          Picasso.HANDLER.post {
+            throw IllegalStateException("Transformation ${transformation.key()} returned a recycled Bitmap.")
+          }
+
+          return null
+        }
+
+        res = newResult
+      }
+
+      return res
+    }
+  }
+
+  val sequence: Int = SEQUENCE_GENERATOR.incrementAndGet()
+
+  var action: Action? = action
+    private set
+
+  private var _actions: MutableList<Action>? = null
+  val actions: List<Action>?
+    get() = _actions
+
+  var future: Future<*>? = null
+  var result: RequestHandler.Result? = null
+    private set
+  var exception: Exception? = null
+    private set
+
+  var priority: Picasso.Priority = priority
+    private set
+
+  val isCancelled: Boolean
+    get() = future?.isCancelled ?: false
+
+  override fun run() {
+    try {
+      updateThreadName(data)
+
+      if (picasso.loggingEnabled) {
+        Utils.log(Utils.OWNER_HUNTER, Utils.VERB_EXECUTING, Utils.getLogIdsForHunter(this))
+      }
+
+      result = hunt()
+      dispatcher.dispatchComplete(this)
+    } catch (e: IOException) {
+      exception = e
+      dispatcher.dispatchRetry(this)
+    } catch (e: Exception) {
+      exception = e
+      dispatcher.dispatchFailed(this)
+    } finally {
+      Thread.currentThread().name = Utils.THREAD_IDLE_NAME
+    }
+  }
+
+  @Throws(IOException::class)
+  fun hunt(): RequestHandler.Result.Bitmap? {
```

Comment:
return non-null type

Reply:
I think return nullable type,  because `applyTransformations()` is return nullable value.  
L236 https://github.com/square/picasso/pull/2141/files#diff-ef14199d5e270d21921bd04ee5049ea2R236

If `hunt()`s return non-null type, will use of  `applyTransformations()!!` ?

Reply:
Ah, fair point, that's unfortunate due to the HANDLER.post.  That can be addressed in a follow-up.